### ISSUE TITLE
driver.execute_script is detectable

### ISSUE THREAD

**Issue by fabifont** [ultrafunkamsterdam/undetected-chromedriver#1017]
Every `execute_script` is detectable by fingerprint js. You can try running `driver.execute_script("let abc = 'abc'")` after getting their page and you will be detected. To avoid this you must run the script directly from the runtime using the cdp protocol.

**Comment by fabifont** [comment]
Simple detection example:

```pyhton3
driver.get("https://fingerprint.com/products/bot-detection/")
driver.execute_script("let abc = 'abc'")
```

**Comment by whhsw** [comment]
I have patched the cdc_* patcher. but looks like not only "excute_script", even find_element triggers detection..

driver = uc.Chrome()
driver.get("https://fingerprint.com/products/bot-detection/")
driver.find_element(By.CLASS_NAME, "Posts-module--name--f89ab")  # this is a element from the page

**Comment by whhsw** [comment]
looking into the source code, find_element is calling driver.find_elements and find_elements is calling driver.execute_script.
The worst news is, execute_script is being used a lot in webdriver.py

**Comment by alexreg** [comment]
@whhsw What if you replace all the `execute_script` calls with `execute_cdp_cmd` calls?

**Comment by whhsw** [comment]
I am not very familiar with the webdriver protocol, was trying to re-write execute_script but not sure how to convert


    def execute_script(self, script, *args):
        """Synchronously Executes JavaScript in the current window/frame.

        :Args:
         - script: The JavaScript to execute.
         - \\*args: Any applicable arguments for your JavaScript.

        :Usage:
            ::

                driver.execute_script('return document.title;')
        """
        if isinstance(script, ScriptKey):
            try:
                script = self.pinned_scripts[script.id]
            except KeyError:
                raise JavascriptException("Pinned script could not be found")

        converted_args = list(args)
        command = Command.W3C_EXECUTE_SCRIPT

        return self.execute(command, {"script": script, "args": converted_args})["value"]

and another one is

    def execute_cdp_cmd(self, cmd: str, cmd_args: dict):
        """Execute Chrome Devtools Protocol command and get returned result The
        command and command args should follow chrome devtools protocol
        domains/commands, refer to link
        https://chromedevtools.github.io/devtools-protocol/

        :Args:
         - cmd: A str, command name
         - cmd_args: A dict, command args. empty dict {} if there is no command args
        :Usage:
            ::
                driver.execute_cdp_cmd('Network.getResponseBody', {'requestId': requestId})
        :Returns:
            A dict, empty dict {} if there is no result to return.
            For example to getResponseBody:
            {'base64Encoded': False, 'body': 'response body string'}
        """
        return self.execute("executeCdpCommand", {"cmd": cmd, "params": cmd_args})["value"]
> @whhsw What if you replace all the `execute_script` calls with `execute_cdp_cmd` calls?

**Comment by alexreg** [comment]
@whhsw I was just playing around with the code myself... I have even worse news. The following line of `find_element` triggers bot detection, even when `execute_script` is not called.
```
return self.execute(Command.FIND_ELEMENT, {"using": by, "value": value})["value"]
```

**Comment by fabifont** [comment]
> @whhsw I was just playing around with the code myself... I have even worse news. The following line of `find_element` triggers bot detection, even when `execute_script` is not called.
> 
> ```
> return self.execute(Command.FIND_ELEMENT, {"using": by, "value": value})["value"]
> ```
> 
> @lukect Hope you don't mind me pinging you... I'm using your fork, and the unfortunately the above is still an issue. Perhaps you have an idea of what sort of detectable trace `Command.FIND_ELEMENT` leaves?

I can confirm that also Command are detectable. Probably they can intercept http requests made from selenium (`execute` make a `request` internally)

**Comment by alexreg** [comment]
@fabifont Yeah, most likely. Looks like it might need another patch of the binary to get around this...

**Comment by fabifont** [comment]
> @fabifont Yeah, most likely. Looks like it might need another patch of the binary to get around this...

I don't think this fixable with a driver patch

**Comment by alexreg** [comment]
> I don't think this fixable with a driver patch

@fabifont To be honest I'm kind of concerned this is a security breach in any case. I wonder if Chrome itself will patch this.

**Comment by lukect** [comment]
Does it not work on all of these Selenium versions: 4.6, 4.7, 4.8?

**Comment by alexreg** [comment]
@lukect I've just tested with Selenium 4.8 (doesn't work).

**Comment by lukect** [comment]
> @lukect I've just tested with Selenium 4.8 (doesn't work).

What about 4.6 and 4.7? I feel this must be something Selenium changed very recently. There's just no way these basic functions have always left behind detectable props and the antibots only figured that out now. 😂 Selenium must've changed somewhere!

**Comment by alexreg** [comment]
@lukect Yeah, it's odd. Seems to fail with Selenium 4.7.2 and 4.6.1 too, however. Maybe some fancy new method that Fingerprint added (and probably other anti-bot systems too)?

**Comment by alexreg** [comment]
@lukect Yeah, it's odd. Seems to fail with Selenium 4.7.2 and 4.6.1 too, however – I just tested.

Maybe some tricky new method that Fingerprint just added (and probably other anti-bot systems too)? But evidently lots of websites don't use it yet, since uc is still working there.

**Comment by fabifont** [comment]
> @lukect Yeah, it's odd. Seems to fail with Selenium 4.7.2 and 4.6.1 too, however. Maybe some fancy new method that Fingerprint added (and probably other anti-bot systems too)?

Yes, detected with every selenium version. Tried both the normal `find` and the relative one.

**Comment by fabifont** [comment]
> > @lukect I've just tested with Selenium 4.8 (doesn't work).
> 
> What about 4.6 and 4.7? I feel this must be something Selenium changed very recently. There's just no way these basic functions have always left behind detectable props and the antibots only figured that out now. joy Selenium must've changed somewhere!

Yes, detected with every selenium version. Tried both the normal find and the relative one.

**Comment by lukect** [comment]
@alexreg @fabifont I'll fix in a few hours if I can reproduce the issue. When exactly are you guys executing a script?

**Comment by fabifont** [comment]
> @alexreg @fabifont I'll fix in a few hours if I can reproduce the issue. When exactly are you guys executing a script?

```python3
driver = uc.Chrome()
driver.get("https://fingerprint.com/products/bot-detection/")
driver.execute_script("let abc = 'abc'")
```

**Comment by fabifont** [comment]
> dont know what the problem is , but given the responses... try this..:
> 
> ```python
> import undetected_chromedriver as uc
> driver = uc.Chrome()
> driver.get('https://www.google.com')   #example 
> cdp = uc.CDP(driver.options)
> await cdp.c.send('Runtime.evaluate' , {'expression': 'alert("hi")'})
> ```

Running JS in the runtime with cdp works as I said above. The problem is that selenium is based entirely on `execute` and `execute_script` calls and those are detected by FinderprintJS BotD

**Comment by alexreg** [comment]
Thanks for your reply, @ultrafunkamsterdam. Yes, I agree exactly with what @fabifont writes. So any call to `find_element` even can be detected, which is kind of impossible to get around some major patching to the driver and/or browser, as I see it.

**Comment by ultrafunkamsterdam** [comment]
# PINPOINTED THE POSSIBLE CAUSE
## since a video says more than 34,5 words.... link below
 
<figure>
 <video src="https://nowsecure.nl/SillyTheory.mp4"></video>
 <figcaption><a href="https://nowsecure.nl/SillyTheory.mp4">DAMN GITHUB. FUCKING ALLOW VIDEO EMBEDING YOU MORONS</a></figcaption>
</figure>

## Link above ( yes that's the correct one

**Comment by alexreg** [comment]
@ultrafunkamsterdam The video doesn't load for me unfortunately. But I'm glad to hear you may have already identified it!

**Comment by fabifont** [comment]
> # PINPOINTED THE POSSIBLE CAUSE
> ## since a video says more than 34,5 words.... link below
> 
> **ps. sorry for the watermark. @work pc don't have Premiere/AE**
> [DAMN GITHUB. FUCKING ALLOW VIDEO EMBEDING YOU MORONS](https://nowsecure.nl/SillyTheory.mp4)
> ## Link above ( yes that's the correct one

Ok, they are goofy asses but, with the new patcher PR what you have shown won't happen anymore since those variables are removed while patching the driver (and If I get the page like you did I won't be detected). The weird thing is that I am getting detected only if I run a `find_element` or `execute_script` before their "detection script" ends. I just wanted to understand what was happening there.

**Comment by lukect** [comment]
> # PINPOINTED THE POSSIBLE CAUSE
> ## since a video says more than 34,5 words.... link below
> 
> **ps. sorry for the watermark. @work pc don't have Premiere/AE**
> [DAMN GITHUB. FUCKING ALLOW VIDEO EMBEDING YOU MORONS](https://nowsecure.nl/SillyTheory.mp4)
> ## Link above ( yes that's the correct one

My PR already fixes this one. The new detections involve something else in addition to this. I will explain tonight.

**Comment by fabifont** [comment]
> > # PINPOINTED THE POSSIBLE CAUSE
> > ## since a video says more than 34,5 words.... link below
> > **ps. sorry for the watermark. @work pc don't have Premiere/AE**
> > [DAMN GITHUB. FUCKING ALLOW VIDEO EMBEDING YOU MORONS](https://nowsecure.nl/SillyTheory.mp4)
> > ## Link above ( yes that's the correct one
> 
> #1010 already fixes this one. The new detections involve something else in addition to this. I will explain tonight.

I also tried the opensource version and I never get detected there (https://fingerprintjs.github.io/BotD/main) .

**Comment by lukect** [comment]
Fixed: #1010

**Comment by fabifont** [comment]
> Fixed: #1010

I can confirm that. Waiting for some small improvements and maybe for the merge.

**Comment by ultrafunkamsterdam** [comment]
Nailed it
